src/v/utils/delta_for.h

```diff
@@ -118,6 +123,119 @@ struct delta_delta {
     bool operator==(delta_delta const&) const = default;
 };
 
+namespace decomp {
+// this namespace contains instruction on how to decompose N_BITS (<=64) into a
+// collection of unsigned types (uint64,32,16,8) and a bunch of residual bits it
+// is used to serialize a collection of values reducing the total space used.
+// everything is done in alittle endian fashion. example: N_BITS=41, get
+// decomposed as a uint32_t for [bits[0], bits[32]), uint8_t for [bits[32],
+// bits[40]) and one residual bit for bits[40]
+
+// detail: an array of 4 (is-unsigned-type-used, number-of-bits-of-this-type)
+// pairs, from uint64_t to uint8_t, that describes if N_BITS will be decomposed
+// with a particular type or not.
+template<size_t N_BITS>
+constexpr auto detail_decomp = [] {
+    auto decomp = std::to_array<std::pair<bool, size_t>>({
+      {false, 8 * sizeof(uint64_t)},
+      {false, 8 * sizeof(uint32_t)},
+      {false, 8 * sizeof(uint16_t)},
+      {false, 8 * sizeof(uint8_t)},
+    });
+
+    auto rem_bits = N_BITS;
+    for (auto& [in_use, sz_bits] : decomp) {
+        in_use = rem_bits >= sz_bits;
+        if (in_use) {
+            rem_bits -= sz_bits;
+        }
+    }
+    return decomp;
+}();
+
+// an array of pairs (bytes-to-save, bit-index-of-this-type-in-decomposition)
+// that describes how to decompose N_BITS and remains only with a residual part
+// < 8 bits
+template<size_t N_BITS>
+constexpr auto unsigned_decomposition = [] {
+    auto decomposition = std::array<
+      std::pair<size_t, size_t>,
+      std::ranges::count_if(
+        detail_decomp<N_BITS>, [](auto& elem) { return elem.first; })>{};
+    auto it = decomposition.begin();
+    size_t acc = 0;
+    for (auto [in_use, sz_bits] : detail_decomp<N_BITS>) {
+        if (in_use) {
+            *(it++) = std::pair{sz_bits / 8, acc};
+            acc += sz_bits;
+        }
+    }
+    return decomposition;
```

Comment:
nit: add static assert that checks that we counted up to N_BITS?

Reply:
did something in the last commit